### PR TITLE
Fix #4798: Incorrect output for object rest destructuring inside array destructuring 

### LINKED ISSUE #4798: Nodes: Incorrect output for object rest destructuring inside array destructuring
I stumbled upon this while preparing tests for #4796.
` [{a, rest...}] = arr ` compiles to  ` [_extends({a}, rest)] = arr `, 
but instead it shoud be something like
```javascript
[ref] = arr; 
({a} = ref);
rest = objectWithoutKeys(ref, ['a']);
```
[Example](https://babeljs.io/repl/#?babili=false&browsers=&build=&builtIns=false&code_lz=G4QwTgBCA0FgpgZwC63GAvBA2gbxAFwCMsARgQEwC-AugNwBQeMEAdOwirRFuow0A&debug=false&circleciRepo=&evaluate=true&lineWrap=false&presets=stage-3%2Cenv&prettier=false&targets=&version=6.26.0) on Babel.

### PR DESCRIPTION
**Pull request by zdenko**
Fixes #4798
```javascript
// [{a, rest...}] = arr
({a} = ref = arr[0]);
rest = objectWithoutKeys(ref, ['a']);

// [b,{a, rest...},c] = arr
b = arr[0], (({a} = ref = arr[1]), rest = objectWithoutKeys(ref, ['a'])), c = arr[2];
```

### CODE REVIEW

File: test/assignment.coffee [hunk 1]
```diff
@@ -185,6 +185,12 @@ test "#4787 destructuring of objects within arrays", ->
   eq b, arr[1].b
   deepEqual {a, b}, arr[1]
 
+test "#4798 destructuring of objects with splat within arrays", ->
+  arr = [1, {a:1, b:2}]
+  [...,{a, r...}] = arr
```
Comment (GeoffreyBooth) [inline test/assignment.coffee]: We should probably also have a test without an expansion.

**Comment by GeoffreyBooth** [comment]
Other than that, it looks good to me. @connec, do you have any notes?

File: test/assignment.coffee [hunk 2]
```diff
@@ -185,6 +185,17 @@ test "#4787 destructuring of objects within arrays", ->
...
+  eq a, 1
+  deepEqual r, {b:2}
+  [b, {q...}] = arr
```
Comment (GeoffreyBooth) [inline test/assignment.coffee]: This reuses `b` from the first test, it should get its own `test "#4798...` block.

Comment (zdenko) [reply test/assignment.coffee]: From the first assignment (`[...,{a, r...}] = arr`)  variables `a` and `r` are set.
The second assignment `[b, {q...}] = arr` sets variables `b` and `q`. 
I think `b` doesn't exist before the second assignment, so it's not reused.

**Comment by zdenko** [comment]
@GeoffreyBooth do I need to add/change something to this PR?

**Comment by GeoffreyBooth** [comment]
Nope, I'm satisfied if @lydell is satisfied.

**Comment by GeoffreyBooth** [comment]
Sorry, I thought this was the other PR. @connec did you have any notes on this?

**Comment by connec** [comment]
LGTM 👍

**Pull request merged**